src/csharp/Microsoft.Spark/Sql/ForeachWriter.cs

```diff
@@ -0,0 +1,190 @@
+        /// </item>
+        /// </list>
+        /// </summary>
+        /// <param name="errorOrNull">
+        /// The <see cref="Exception"/> thrown during processing or null if there was no error.
+        /// </param>
+        void Close(Exception errorOrNull);
+    }
+
+    /// <summary>
+    /// Wraps a <see cref="IForeachWriter"/> and calls the appropriate methods as decribed in
+    /// the lifecycle documentation for the interface.
+    /// </summary>
+    internal class ForeachWriterWrapper
+    {
+        private readonly IForeachWriter _foreachWriter;
+
+        internal ForeachWriterWrapper(IForeachWriter foreachWriter) =>
+            _foreachWriter = foreachWriter;
+
+        internal IEnumerable<object> Execute(int partitionId, IEnumerable<Row> rows)
+        {
+            if (!TaskContextHolder.Get().LocalProperties.TryGetValue(
+                "streaming.sql.batchId",
+                out string epochIdStr) || !long.TryParse(epochIdStr, out long epochId))
+            {
+                throw new Exception(
+                    $"Could not get or parse batch id from TaskContext - batchId: {epochIdStr}");
+            }
+
+            Exception error = null;
+            bool opened = _foreachWriter.Open(partitionId, epochId);
+
+            try
+            {
+                if (opened)
+                {
+                    foreach (Row row in rows)
+                    {
+                        _foreachWriter.Process(row);
+                    }
+                }
+            }
+            catch (Exception e)
+            {
+                error = e;
+            }
+            finally
+            {
+                _foreachWriter.Close(error);
+
+                if (error != null)
+                {
+                    throw error;
+                }
+            }
+
+            return Enumerable.Empty<object>();
+        }
+    }
+
+    /// <summary>
+    /// Wraps the given Func object, which represents a <see cref="ForeachWriterWrapper"/> UDF.
+    /// When this UdfWrapper is processed, the PythonEvalType is
+    /// <see cref="Utils.UdfUtils.PythonEvalType.NON_UDF"/>. The CommandExecutor expects the 
+    /// <see cref="Execute(int, IEnumerable{object})"/> method to match the
+    /// <see cref="RDD.WorkerFunction.ExecuteDelegate"/> delegate. This UdfWrapper helps map 
+    /// the UDF for <see cref="ForeachWriterWrapper"/> to <see cref="RDD.WorkerFunction.ExecuteDelegate"/>.
+    /// </summary>
+    [UdfWrapper]
+    internal class ForeachWriterWrapperUdfWrapper
+    {
+        private readonly Func<int, IEnumerable<Row>, IEnumerable<object>> _func;
+
+        internal ForeachWriterWrapperUdfWrapper(
+            Func<int, IEnumerable<Row>, IEnumerable<object>> func)
+        {
+            _func = func;
+        }
+
+        internal IEnumerable<object> Execute(int pid, IEnumerable<object> input)
+        {
+            return _func(
+                pid,
+                input.Cast<object[]>().SelectMany(
```

Comment:
Would cast here will materialize the `IEnumerable` to array or no?

Reply:
Cast will call `CastIterator<RowConstructor>`, which returns an `IEnumerable<RowConstructor>`.  The casting won't occur unless the IEnumerable is materialized to an array or the elements are individually process through some foreach call.

For more details:
- https://github.com/microsoft/referencesource/blob/17b97365645da62cf8a49444d979f94a59bbb155/System.Core/System/Linq/Enumerable.cs#L1031-L1040
- https://docs.microsoft.com/en-us/dotnet/api/system.linq.enumerable.cast?view=netframework-4.8#remarks

Reply:
ok. also, you expect the following right? the size of `object[]` is 1? If so, can we be explicitly about it?

Reply:
The linq query doesn't make any assumptions regarding the size of `object[]`. It should work if `object[]` is 1 or more, as long as the contents of the `object`s within the `object[]` are `RowConstructor`'s

Reply:
it's not about linq. you expect one row in your udf, no?

Reply:
I don't think generalizing the solution is a good idea here.

Reply:
The `ForeachWriterWrapper.Execute()` (_func in this case) expects an `IEnumerable<Row>`

Reply:
right, so object[] should have one element, so I would assert here to be explicit.

Reply:
RowConstructor.GetRow() is now called from the `RowDeserializer` with a comment regarding `AutoBatchedPickler` which determines number of Rows being pickled.

Reply:
Also added comment to `ForeachWriterWrapperUdfWrapper.Execute` which also mentions `AutoBatchedPickler` as well as its use in the `PythonForeachWriter.scala` class.

Reply:
Yea, now it looks much better, thanks!